File: DaS.ScriptLib.CS/Game/Data/Structures/EntityAIController.vb
Imports DaS.ScriptLib.Injection

Namespace Game.Data.Structures

    Public Class EntityAIController

        Public ReadOnly Property Address As Integer
            Get
                Return AddressReadFunc()
            End Get
        End Property

        Public AddressReadFunc As Func(Of Integer)

        Public Sub New(ptr As Integer)
            Dim ptrDeref As Integer = ptr
            AddressReadFunc = Function() ptrDeref
        End Sub

        Public Sub CopyFrom(other As EntityAIController)
            'TODO
        End Sub

        Public Property EntityPtr As Integer
            Get
                Return RInt32(Address + &H14)
            End Get
            Set(value As Integer)
                WInt32(Address + &H14, value)
            End Set
        End Property

        Public ReadOnly Property Entity As Entity
            Get
                Return New Entity(Function() EntityPtr)
            End Get
        End Property

        Public Property AIScript As Integer
            Get
                Return RInt32(Address + &H78)
            End Get
            Set(value As Integer)
                WInt32(Address + &H78, value)
            End Set
        End Property

        Public Property AnimationID As Integer
            Get
                Return RInt32(Address + &H9C)
            End Get
            Set(value As Integer)
                WInt32(Address + &H9C, value)
            End Set
        End Property

        Public Property AIScript2 As Integer
            Get
                Return RInt32(Address + &H80)
            End Get
            Set(value As Integer)
                WInt32(Address + &H80, value)
            End Set
        End Property

        Public Property PosX As Single
            Get
                Return RFloat(Address + &H1E0)
            End Get
            Set(value As Single)
                WFloat(Address + &H1E0, value)
            End Set
        End Property

        Public Property PosY As Single
            Get
                Return RFloat(Address + &H1E4)
            End Get
            Set(value As Single)
                WFloat(Address + &H1E4, value)
            End Set
        End Property

        Public Property PosZ As Single
            Get
                Return RFloat(Address + &H1E8)
            End Get
            Set(value As Single)
                WFloat(Address + &H1E8, value)
            End Set
        End Property

        Public Property RotZ As Single
            Get
                Return RFloat(Address + &H1E8)
            End Get
            Set(value As Single)
                WFloat(Address + &H1E8, value)
            End Set
        End Property

        Public Property RotY As Single
            Get
                Return RFloat(Address + &H1F0)
            End Get
            Set(value As Single)
                WFloat(Address + &H1F0, value)
            End Set
        End Property

        Public Property RotX As Single
            Get
                Return RFloat(Address + &H1F4)
            End Get
            Set(value As Single)
                WFloat(Address + &H1F4, value)
            End Set
        End Property

        Public Property AnimationID2 As Integer
            Get
                Return RInt32(Address + &H208)
            End Get
            Set(value As Integer)
                WInt32(Address + &H208, value)
            End Set
        End Property

    End Class

End Namespace

File: DaS.ScriptLib.NEW/Game/Data/Structures/EntityAIController.cs
﻿using System;
using DaS.ScriptLib.Injection;

namespace DaS.ScriptLib.Game.Data.Structures
{

    public class EntityAIController
	{

		public int Address {
			get { return AddressReadFunc(); }
		}


		public Func<int> AddressReadFunc;
		public EntityAIController(int ptr)
		{
			int ptrDeref = ptr;
			AddressReadFunc = () => ptrDeref;
		}

		public void CopyFrom(EntityAIController other)
		{
			//TODO
		}

		public int EntityPtr {
			get { return Hook.RInt32(Address + 0x14); }
			set { Hook.WInt32(Address + 0x14, value); }
		}

		public Entity Entity {
			get { return new Entity(() => EntityPtr); }
		}

		public int AIScript {
			get { return Hook.RInt32(Address + 0x78); }
			set { Hook.WInt32(Address + 0x78, value); }
		}

		public int AnimationID {
			get { return Hook.RInt32(Address + 0x9c); }
			set { Hook.WInt32(Address + 0x9c, value); }
		}

		public int AIScript2 {
			get { return Hook.RInt32(Address + 0x80); }
			set { Hook.WInt32(Address + 0x80, value); }
		}

		public float PosX {
			get { return Hook.RFloat(Address + 0x1e0); }
			set { Hook.WFloat(Address + 0x1e0, value); }
		}

		public float PosY {
			get { return Hook.RFloat(Address + 0x1e4); }
			set { Hook.WFloat(Address + 0x1e4, value); }
		}

		public float PosZ {
			get { return Hook.RFloat(Address + 0x1e8); }
			set { Hook.WFloat(Address + 0x1e8, value); }
		}

		public float RotZ {
			get { return Hook.RFloat(Address + 0x1e8); }
			set { Hook.WFloat(Address + 0x1e8, value); }
		}

		public float RotY {
			get { return Hook.RFloat(Address + 0x1f0); }
			set { Hook.WFloat(Address + 0x1f0, value); }
		}

		public float RotX {
			get { return Hook.RFloat(Address + 0x1f4); }
			set { Hook.WFloat(Address + 0x1f4, value); }
		}

		public int AnimationID2 {
			get { return Hook.RInt32(Address + 0x208); }
			set { Hook.WInt32(Address + 0x208, value); }
		}

	}

}